04-cs-fundamentals/internship/graphs.md

```diff
@@ -0,0 +1,421 @@
+
+```ruby
+adjacency_list_with_hash = {
+  "Seattle" => ["Los Angelos", "Topeka"],
+  "Los Angeles" => ["Seattle", "Denver", "San Diego"],
+  "Topeka" => ["Seattle"],
+  "Denver" => ["Los Angeles", "Dallas"],
+  "San Diego" => ["Los Angeles"],
+  "Dallas" => ["Denver"]
+}
+```
+
+
+![Adjacency list diagram with a hash table](images/graphs-adjacency-list-cities.png)
+
+This representation provides similar benefits to using an array for an adjacency list, but provides us more flexibility with keys.  We no longer need numeric ids for our nodes and the ids do not need to be consecutive.  For example, in the adjacency matrix and list above, there was no node 0, but we maintained space for a node with that id.
+
+## Graph Algorithms
+
+There are a huge [class of problems](https://en.wikipedia.org/wiki/Category:Graph_algorithms) involving graphs.  We will look at a few as a sample.  
+
+### Breadth-First-Search
+
+Like in a Binary Search Tree a breadth-first-search in a graph is performed with a queue.  However because a graph can, and likely will have a cycle, unlike a tree we need to keep track of the nodes we have visited.  
+
+In breadth-first-search we start with a particular node and visit each node connected to the starting point in the graph starting with the closest node to the starting point and expanding outward.
+
+We do so by adding each of the neighbors of the starting node to a Queue and then loop through the Queue removing an element and repeating the process with the neighbors of the current node.
+
+**Pseudocode**
+
+1. Start with a graph adjacency list and a specific node called `start_node`
+1. Create a list of nodes called `visited` and mark each element `false`
+1. Create an empty queue `q`
+1. Add `start_node` to `q`
+1. While `q` is not empty
+    1. Remove an element from `q` and store it in `current`
+    1. Mark `current` as `true` in the list `visited`
+    1. Loop through each of `current`'s neighbors
+        1. If the neighbor is not `true` in `visited` add it to `q`
+
+You can see breadth-first-search animated on [HackerEarth](https://www.hackerearth.com/practice/algorithms/graphs/breadth-first-search/visualize/)
+
+Breadth first search is a solution in a variety of problems including:
+
+- Finding the shortest path in a graph/maze
+- Solving puzzle games like a [Rubik's Cube](https://www.quora.com/How-can-solving-a-Rubiks-Cube-be-framed-as-a-graph-problem)
+- Checking to see if a graph is connected
+
+**Questions**
+
+<details>
+  <summary>What is the time complexity of Breadth-First-Search</summary>
+  Since you will visit each node once, and loop through each of the edges in each node the Big-O of this algorithm is O(N + E) where `N` is the number of nodes in the graph and `E` is the number of edges since each node and each edge will be explored.
```

Comment:
```suggestion
  Since you will visit each node once, and loop through each of the edges in each node the Big-O of this algorithm is _O(N + E)_ where _N_ is the number of nodes in the graph and _E_ is the number of edges since each node and each edge will be explored.
```